bundles/core/org.eclipse.smarthome.core.audio/src/main/java/org/eclipse/smarthome/core/audio/AudioManager.java

```diff
@@ -0,0 +1,460 @@
+            try {
+                return sink.getVolume();
+            } catch (IOException e) {
+                logger.error("An exception occured while getting the volume of sink {} : '{}'", sink.getId(),
+                        e.getMessage());
+            }
+        }
+
+        return 0;
+
+    }
+
+    /**
+     * Sets the volume for a sink.
+     *
+     * @param sinkId the sink to set the volume
+     * @param volume the volume to set as a value between 0 and 1
+     */
+    public void setVolume(String sinkId, float volume) {
+        AudioSink sink = null;
+        if (sinkId == null) {
+            sink = getSink();
+        } else {
+            sink = audioSinks.get(sinkId);
+        }
+
+        if (sink != null) {
+            try {
+                sink.setVolume(volume);
+            } catch (IOException e) {
+                logger.error("An exception occured while setting the volume of sink {} : '{}'", sink.getId(),
+                        e.getMessage());
+            }
+        }
+    }
+
+    /**
+     * Retrieves an AudioSource.
+     * If a default name is configured and the service available, this is returned. Otherwise, the first available
+     * service is returned.
+     *
+     * @return an AudioSource or null, if no service is available or if a default is configured, but no according
+     *         service is found
+     */
+    public AudioSource getSource() {
+        AudioSource source = null;
+        if (defaultSource != null) {
+            source = audioSources.get(defaultSource);
+            if (source == null) {
+                logger.warn("Default AudioSource service '{}' not available!", defaultSource);
+            }
+        } else if (!audioSources.isEmpty()) {
+            source = audioSources.values().iterator().next();
+        } else {
+            logger.debug("No AudioSource service available!");
+        }
+        return source;
+    }
+
+    /**
+     * Retrieves an AudioSink.
+     * If a default name is configured and the service available, this is returned. Otherwise, the first available
+     * service is returned.
+     *
+     * @return an AudioSink or null, if no service is available or if a default is configured, but no according service
+     *         is found
+     */
+    public AudioSink getSink() {
+        AudioSink sink = null;
+        if (defaultSink != null) {
+            sink = audioSinks.get(defaultSink);
+            if (sink == null) {
+                logger.warn("Default AudioSink service '{}' not available!", defaultSink);
+            }
+        } else if (!audioSinks.isEmpty()) {
+            sink = audioSinks.values().iterator().next();
+        } else {
+            logger.debug("No AudioSink service available!");
+        }
+        return sink;
+    }
+
+    /**
+     * Retrieves the ids of all sources
+     *
+     * @return ids of all sources
+     */
+    public Set<String> getSourceIds() {
+        return audioSources.keySet();
```

Comment:
The keySet is a view... So, the caller could run into a CME using the returned Set it the Map changes.
Wouldn't it be better to return a new set created by the content of the view set?